py/label_microservice/repo_specific_model.py

```diff
@@ -0,0 +1,167 @@
+
+    Args:
+      repo_owner: str
+      repo_name: str
+      embedding_api_endpoint: The endpoint for the microservice to compute
+        embeddings
+    """
+
+    model = RepoSpecificLabelModel()
+
+    model._embedding_api_endpoint = embedding_api_endpoint
+
+    model.config = repo_config.RepoConfig(repo_owner=repo_owner,
+                                          repo_name=repo_name)
+
+    # download model
+    gcs_util.download_file_from_gcs(model.config.model_bucket_name,
+                                    model.config.model_gcs_path,
+                                    model.config.model_local_path)
+
+    # download label columns
+    gcs_util.download_file_from_gcs(model.config.model_bucket_name,
+                                    model.config.labels_gcs_path,
+                                    model.config.labels_local_path)
+
+    model._mlp_predictor = mlp.MLPWrapper(
+      clf=None, model_file=model.config.model_local_path, load_from_model=True)
+
+    # Get label info.
+    # Expect a YAML file with a dictionary
+    # {'labels': list, 'probability_thresholds': {label_index: threshold}}
+    with open(model.config.labels_local_path, 'r') as f:
+      label_columns = yaml.safe_load(f)
+
+    model._label_names = label_columns["labels"]
+    model._label_thresholds = {}
+
+    for index, threshold in label_columns["probability_thresholds"].items():
+      model._label_thresholds[model._label_names[index]] = threshold
+
+    logging.info(f"Loaded model gs://{model.config.model_bucket_name}/"
+                 f"{model.config.model_gcs_path}")
+    logging.info(f"Loaded model config gs://{model.config.model_bucket_name}/"
+                 f"{model.config.labels_gcs_path}")
+    logging.info(f"Model label thresholds {model._label_thresholds}")
+    if not embedding_api_endpoint:
+      embedding_api_endpoint = DEFAULT_EMBEDDING_API_ENDPOINT
+
+    model._embedding_api_endpoint = embedding_api_endpoint
+    logging.info(f"Issue embedding service set to {model._embedding_api_endpoint}")
+    return model
+
+  def predict_issue_labels(self, title:str , text:str):
+    """Return a dictionary of label probabilities.
+
+    Args:
+      title: The title for the issue
+      text: The text for the issue
+
+    Return
+    ------
+    dict: Dictionary of label to probability of that label for the
+      the issue str -> float
+    """
+    issue_embedding = self._get_issue_embedding(title, text)
+
+    # if not retrieve the embedding, ignore to predict it
+    if issue_embedding is None:
+      logging.error("No embeddings returned for issue")
+      return {}
+
+    # change embedding from 1d to 2d for prediction and extract the result
+    label_probabilities = self._mlp_predictor.predict_probabilities(
+      [issue_embedding])[0]
+
+    # check thresholds to get labels that need to be predicted
+    predictions = dict(zip(self._label_names, label_probabilities))
```

Comment:
This makes me a bit uneasy, is there something more we can do to ensure the order of the label names does match the order of the probabilities?

Reply:
Good point. I filed kubeflow/code-intelligence#83 to store the class labels in the MLPClassifier. I'm not sure what else we could do to ensure they stay in sync.

Reply:
👍